### PR TITLE
Iterate on the backend API for device mgmt

### PR DESCRIPTION
**Pull request by wconstab**
Fixes #{issue number}

### CODE REVIEW

File: lazy_tensor_core/lazy_tensor_core/csrc/compiler/backend_impl_interface.h [hunk 1]
```diff
@@ -70,22 +71,31 @@ class BackendImplInterface {
...
+  // Set or get the default device type.
+  // For backends used with virtual c10:: Devices, this configures what real device type
+  // the backend should use, and matters if the backend supports more than one type of real
+  // device.
+  virtual torch_lazy_tensors::BackendDeviceType GetDefaultDeviceType() const = 0;
+  virtual void SetDefaultDeviceType(std::string device_type) const = 0;
 
-  virtual size_t GetNumDevices() const = 0;
+  // Query all available backend devices
+  virtual torch_lazy_tensors::BackendDevice GetBackendDevices() const = 0;
```
Comment (JackCaoG) [inline lazy_tensor_core/lazy_tensor_core/csrc/compiler/backend_impl_interface.h:82]: Shouldn't this return a vector of devices?

Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/compiler/backend_impl_interface.h:82]: ah.. yea!

File: lazy_tensor_core/lazy_tensor_core/csrc/compiler/backend_impl_interface.h [hunk 2]
```diff
@@ -70,22 +71,31 @@ class BackendImplInterface {
    * Device Configuration
    * */
 
-  virtual std::string GetDefaultDevice() const = 0;
+  // Set or get the default device type.
+  // For backends used with virtual c10:: Devices, this configures what real device type
+  // the backend should use, and matters if the backend supports more than one type of real
+  // device.
+  virtual torch_lazy_tensors::BackendDeviceType GetDefaultDeviceType() const = 0;
+  virtual void SetDefaultDeviceType(std::string device_type) const = 0;
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/compiler/backend_impl_interface.h:79]: I think 'DefaultDeviceType' might be an unclear name.

What about {Set/Get}VirtualDeviceHardwareType?  

The point is, this DeviceType is more than just a default; it's the type we always use when you use our virtual device.  It's fine that there is a 'default' HardwareType, and that can be returned by GetVirtualDeviceHardwareType by default.

I'm not sure what this implies about using multiple kinds of hardware in the same program.  (A) not sure we want to support this, (B) if we do support it, is it just done by calling Set multiple times in the program?

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp [hunk 3]
```diff
@@ -233,8 +232,10 @@ LazyNativeFunctions::convolution_backward_overrideable(
           << " input=" << input.toString() << " weight=" << weight.toString();
   const auto kernel_size = weight.sizes().slice(2);
   CHECK(kernel_size.size() == 2 || kernel_size.size() == 3);
+  // TODO(whc) make a backend API for 'getEagerFallbackDevice(lazy::Device)'?
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp:235]: delete this todo

**Comment by wconstab** [comment]
@JackCaoG I'm planning to land this as-is, to unblock some other stuff.  But we can still change it to satisfy your requirements for default device or other things you discover..

**Review by desertfire — APPROVED** [review]
Let's roll it!

**Pull request merged**